securedrop/journalist_app/api.py

```diff
@@ -124,6 +132,50 @@ def get_all_sources() -> Tuple[flask.Response, int]:
         sources = Source.query.filter_by(pending=False, deleted_at=None).all()
         return jsonify({"sources": [source.to_json() for source in sources]}), 200
 
+    @api.route("/head", methods=["GET"])
+    def head(version: Optional[str] = None) -> Tuple[flask.Response, int]:
+        try:
+            current = redis_client.get("version").decode()
+        except AttributeError:
+            sources = Source.query.filter_by(pending=False, deleted_at=None).all()
+            submissions = Submission.query.all()
+            replies = Reply.query.all()
+            index = {
+                "sources": {source.uuid: source.version for source in sources},
+            }
+            current = json_version(index)
+            redis_client.set("version", current, ex=HOUR)
+
+        if request.headers.get("If-None-Match") == current:
+            return jsonify({}), 304
+
+        # TODO: DRY
+        sources = Source.query.filter_by(pending=False, deleted_at=None).all()
+        submissions = Submission.query.all()
+        replies = Reply.query.all()
+        index = {
+            "sources": {source.uuid: source.version for source in sources},
+        }
+        response = jsonify(index)
+        response.headers["ETag"] = current
+        return response, 200
+
+    @api.route("/index", methods=["POST"])
```

Comment:
nit: this could be a `GET /sources`, since we're just doing a multi-fetch of sources?